### PR TITLE
Add terraform-lsp integration

### PR DESCRIPTION
**Pull request by OJFord**
See https://github.com/juliosueiras/terraform-lsp.

Related: #2726

### CODE REVIEW

**Comment by OJFord** [comment]
I did see the note about testing, but as far as I can tell the tests are more applicable to non-LSP linters/formatters?

If there's an example of an existing LSP integration with meaningful tests I'd be happy to give it a go, but as is probably clear I'm not that comfortable with vimscript!

**Comment by AubreySLavigne** [comment]
I'd like to see this LSP added, so I went ahead and added the tests. This is my first foray into Vim Script, so I'm very open to any feedback you have.

I'm also pretty new to open source contributing -- since I'm not the person who opened the original pull request, should I create a new pull instead?  Other than that, is there anything you'd like to see me to from here?

**Comment by OJFord** [comment]
Thanks @AubreySLavigne! I'm sorry I hadn't gotten to it yet.

> I'm also pretty new to open source contributing -- since I'm not the person who opened the original pull request, should I create a new pull instead?

That's certainly one way we could do it, and I think necessary (on GitHub) when the original author abandons it. But since I'm here, I can just `git cherry-pick` your commit with the tests so that it shows up here, which will keep the history & conversation in one place.

**Comment by AubreySLavigne** [comment]
@OJFord 

I thought I had cleaned up everything, but the Travis CI build is failing with this request.  I've  pushed up 9de308ecb4742035dd90302e95da825fe989c4ad in my fork, so if you include or squash in this commit, the linting step should pass.

I attempted to create a PR from my branch into yours, but I can't find yours in the list for Base Repositories.

File: doc/ale-terraform.txt [hunk 1]
```diff
@@ -32,6 +32,27 @@ g:ale_terraform_terraform_executable     *g:ale_terraform_terraform_executable*
   This variable can be changed to use a different executable for terraform.
 
 
+===============================================================================
+terraform-lsp                                     *ale-terraform-terraform-lsp*
+
+g:ale_terraform_langserver_executable
+                                    *g:ale_ale_terraform_langserver_executable*
```
Comment (cliffxuan) [inline doc/ale-terraform.txt:39]: `g:ale_ale_terraform_langserver_executable` -> `g:ale_terraform_langserver_executable`?

File: doc/ale-terraform.txt [hunk 2]
```diff
@@ -32,6 +32,27 @@ g:ale_terraform_terraform_executable     *g:ale_terraform_terraform_executable*
...
+                                    *b:ale_ale_terraform_langserver_executable*
```
Comment (cliffxuan) [inline doc/ale-terraform.txt:40]: `b:ale_ale_terraform_langserver_executable` -> `b:ale_terraform_langserver_executable`?

**Comment by OJFord** [comment]
The Travis lint job appeared to pass ('OK'), but then says 'Something went wrong' and exited 1. It may just be one-off, but I seem not to have permission to retry.

**Comment by w0rp** [comment]
```
Badly aligned tags follow:

doc/ale-terraform.txt:                                    *g:ale_terraform_langserver_executable*
doc/ale-terraform.txt:                                    *b:ale_terraform_langserver_executable*
```

The lint job is complaining about tags not being aligned to the right margin.

**Comment by OJFord** [comment]
Oh I see, sorry I missed that (I'm used to a test runner that shows `F` instead of `.` when a test fails, so I read `..... Ran 5 tests OK` as all ok).

Thanks for the fix.

**Comment by w0rp** [comment]
Cheers! :beers:

**Comment by stefansedich** [comment]
I am a bit lost here, I don't see terraform-lsp as an available linter, I set it up but does not seem to be working, am I missing something obvious?

```
let g:ale_linters = {'hcl': ['terraform-lsp']}
```

**Comment by OJFord** [comment]
@stefansedich Not obvious, you're right that the language is actually HCL (Hashicorp Configuration Language), but the ale linter is defined for `'terraform'`, which you'll note is the `:set filetype` when you have a .tf file open.

**Pull request merged**